Add CUDA 12.4 workflows

nWEIdia opened the pull request:
Reference: https://github.com/pytorch/pytorch/pull/98492 

cc @albanD @ptrblck @atalman @malfet

nWEIdia committed:
Add CUDA 12.4 workflows

Add cu124 pip dependencies

CU124 Dependency version fix and addressing review comments

Bug fix filtering out cuXXX from arch_version

Aidyn-A commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -367,5 +367,6 @@ def generate_wheels_matrix(
     return ret
 
 
+validate_nccl_dep_consistency("12.2")
```
Is it a typo or it should be indeed `12.2`?

nWEIdia replied:
Good catch Aidyn, it is a typo, thank you!

nWEIdia replied:
Done.

ptrblck commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -58,6 +58,21 @@
+        "nvidia-curand-cu12==10.3.4.107; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-cusolver-cu12==11.6.2.20; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-cusparse-cu12==12.3.0.142; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-nccl-cu12==2.20.3; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```
`nvidia-nccl-cu12==2.20.3` does not exist. You might want to pin it to `==2.20.5`: https://pypi.org/project/nvidia-nccl-cu12/2.20.5/

nWEIdia replied:
Done

ptrblck commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -58,6 +58,21 @@
+        "nvidia-cuda-nvrtc-cu12==12.3.107; platform_system == 'Linux' and platform_machine == 'x86_64' | " 
+        "nvidia-cuda-runtime-cu12==12.4.99; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-cuda-cupti-cu12==12.4.99; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-cudnn-cu12==9.0.0.322; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```
`nvidia-cudnn-cu12==9.0.0.322` does not exist, see: https://github.com/pytorch/pytorch/pull/120642#issuecomment-1974017028

nWEIdia replied:
Using 8.9.7.29 same as what is used in https://github.com/pytorch/pytorch/pull/120642 thanks for the pointer!

ptrblck commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -58,6 +58,21 @@
+        "nvidia-nccl-cu12==2.20.3; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-nvtx-cu12==12.4.99; platform_system == 'Linux' and platform_machine == 'x86_64'"
+        "nvidia-nvjitlink-cu12==12.4.99; platform_system == 'Linux' and platform_machine == 'x86_64'"
+        "nvidia-dali-cuda120==1.35.0; platform_system == 'Linux' and platform_machine == 'x86_64'"
```
Remove DALI

nWEIdia replied:
Done

ptrblck commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -13,13 +13,13 @@
 
 
-CUDA_ARCHES_CUDNN_VERSION = {"11.8": "8", "12.1": "8"}
+CUDA_ARCHES_CUDNN_VERSION = {"11.8": "8", "12.1": "8", "12.4":"9"}
```
Drop cudnn to 8

nWEIdia replied:
Yes

johnnynunez commented:
@ptrblck @nWEIdia why not with cudnn 9.0? It's major improve on flash attention, and I saw that headers were updated to cudnn 9.0

ptrblck commented:
@johnnynunez It's not available as described in the review: https://github.com/pytorch/pytorch/pull/121684#discussion_r1521773989

johnnynunez commented:
> @johnnynunez It's not available as described in the review: [#121684 (comment)](https://github.com/pytorch/pytorch/pull/121684#discussion_r1521773989)

I see.. thanks

nWEIdia commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -58,6 +58,20 @@
+        "nvidia-curand-cu12==10.3.5.119; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-cusolver-cu12==11.6.0.99; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-cusparse-cu12==12.3.0.142; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-nccl-cu12==2.20.5; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```
Good point. NCCL was updated and merged via https://github.com/pytorch/pytorch/pull/121635

johnnynunez commented:
when will it merged? :)

johnnynunez commented:
12.4 update 1 is out:
https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/

johnnynunez commented:
@ptrblck @nWEIdia nvidia cudnn9 now is available nvidia-cudnn-cu12 9.0.0.312
https://pypi.org/project/nvidia-cudnn-cu12/

nWEIdia committed:
Checkout nccl submodule and then run .github//scripts/generate_ci_workflows.py
This auto generates/updates the generated-*.yml files.

atalman commented:
@nWEIdia while windows AMI is not yet in place we would need to add only Linux part of things

atalman commented on .github/workflows/generated-linux-aarch64-binary-manywheel-nightly.yml:
```diff
@@ -54,7 +54,6 @@ jobs:
       ALPINE_IMAGE: "arm64v8/alpine"
       build_name: manywheel-py3_8-cpu-aarch64
       build_environment: linux-aarch64-binary-manywheel
```
This is wrong we need the PYTORCH_EXTRA_INSTALL_REQUIREMENTS here. We want metadata for all the wheels be identical.

atalman commented on .github/workflows/generated-macos-arm64-binary-wheel-nightly.yml:
```diff
@@ -46,7 +46,6 @@ jobs:
       GPU_ARCH_TYPE: cpu
       SKIP_ALL_TESTS: 1
       DESIRED_PYTHON: "3.8"
```
This is wrong we need the PYTORCH_EXTRA_INSTALL_REQUIREMENTS here. We want metadata for all the wheels be identical.

atalman commented on .github/workflows/generated-windows-binary-wheel-nightly.yml:
```diff
@@ -46,7 +46,6 @@ jobs:
       GPU_ARCH_TYPE: cpu
       SKIP_ALL_TESTS: 1
       DESIRED_PYTHON: "3.8"
```
This is wrong we need the PYTORCH_EXTRA_INSTALL_REQUIREMENTS here. We want metadata for all the wheels be identical.

atalman commented on .github/workflows/generated-macos-arm64-binary-wheel-nightly.yml:
```diff
@@ -46,7 +46,6 @@ jobs:
       GPU_ARCH_TYPE: cpu
       SKIP_ALL_TESTS: 1
       DESIRED_PYTHON: "3.8"
-      PYTORCH_EXTRA_INSTALL_REQUIREMENTS: nvidia-cuda-nvrtc-cu12==12.1.105; platform_system == 'Linux' and platform_machine == 'x86_64' | nvidia-cuda-runtime-cu12==12.1.105; platform_system == 'Linux' and platform_machine == 'x86_64' | nvidia-cuda-cupti-cu12==12.1.105; platform_system == 'Linux' and platform_machine == 'x86_64' | nvidia-cudnn-cu12==8.9.2.26; platform_system == 'Linux' and platfor ...
```
This is wrong we need the PYTORCH_EXTRA_INSTALL_REQUIREMENTS here. We want metadata for all the wheels be identical.

nWEIdia replied:
I put the [12.4] back.

atalman commented on .github/workflows/generated-windows-binary-wheel-nightly.yml:
```diff
@@ -46,7 +46,6 @@ jobs:
       GPU_ARCH_TYPE: cpu
       SKIP_ALL_TESTS: 1
       DESIRED_PYTHON: "3.8"
-      PYTORCH_EXTRA_INSTALL_REQUIREMENTS: nvidia-cuda-nvrtc-cu12==12.1.105; platform_system == 'Linux' and platform_machine == 'x86_64' | nvidia-cuda-runtime-cu12==12.1.105; platform_system == 'Linux' and platform_machine == 'x86_64' | nvidia-cuda-cupti-cu12==12.1.105; platform_system == 'Linux' and platform_machine == 'x86_64' | nvidia-cudnn-cu12==8.9.2.26; platform_system == 'Linux' and platfor ...
```
This is wrong we need the PYTORCH_EXTRA_INSTALL_REQUIREMENTS here. We want metadata for all the wheels be identical.

nWEIdia replied:
I put it back

atalman commented on .github/workflows/generated-linux-aarch64-binary-manywheel-nightly.yml:
```diff
@@ -54,7 +54,6 @@ jobs:
       ALPINE_IMAGE: "arm64v8/alpine"
       build_name: manywheel-py3_8-cpu-aarch64
       build_environment: linux-aarch64-binary-manywheel
-      PYTORCH_EXTRA_INSTALL_REQUIREMENTS: nvidia-cuda-nvrtc-cu12==12.1.105; platform_system == 'Linux' and platform_machine == 'x86_64' | nvidia-cuda-runtime-cu12==12.1.105; platform_system == 'Linux' and platform_machine == 'x86_64' | nvidia-cuda-cupti-cu12==12.1.105; platform_system == 'Linux' and platform_machine == 'x86_64' | nvidia-cudnn-cu12==8.9.2.26; platform_system == 'Linux' and platfor ...
```
This is wrong we need the PYTORCH_EXTRA_INSTALL_REQUIREMENTS here. We want metadata for all the wheels be identical.

nWEIdia replied:
Fixed

nWEIdia committed:
Fix dependency string typo

nWEIdia committed:
Disable CUDA 12.4 build for Windows

nWEIdia committed:
Undo PYTORCH_EXTRA_INSTALL_REQUIREMENTS changes

nWEIdia committed:
Test builder patch for conda build failures due to libcupti

nWEIdia committed:
Forgot to run ./regenerate.sh

nWEIdia commented:
Hi @malfet, for pip wheels, the build seems successful, but the test job failed: installing it required the presence of cu124/ AWS directory. https://github.com/pytorch/pytorch/actions/runs/8701283769/job/23866760262   
Could you please help prepare the cu124 directory? https://[download.pytorch.org/whl/nightly/cu124](https://download.pytorch.org/whl/nightly/cu124)  Thanks!

nWEIdia committed:
Revert "Forgot to run ./regenerate.sh"

This reverts commit d43bc6df2d065a52cef8edc4d25986f417e2313a.

nWEIdia committed:
Revert "Test builder patch for conda build failures due to libcupti"

This reverts commit 9254625b00bcd22f214a03d15b1dea6a47c8e172.

nWEIdia committed:
Change builder repo/branch to test this PR and https://github.com/pytorch/builder/pull/1786

nWEIdia committed:
update driver for cu124
reference: https://docs.nvidia.com/cuda/archive/12.4.0/cuda-toolkit-release-notes/index.html#id6

                 Linux x86_64 Driver Version | Windows x86_64 Driver Version

CUDA 12.4 GA        >=550.54.14              |     >=551.61

nWEIdia committed:
Update driver to 550.54.15 according to https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-550-54-15/index.html

nWEIdia committed:
Revert "Update driver to 550.54.15 according to https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-550-54-15/index.html"

This reverts commit 7ec3934759839a40650bacf75560626442a3424f.

nWEIdia committed:
Revert "update driver for cu124"

This reverts commit aaa24b9373e45d918e0c882188a68f0e762b5544.

nWEIdia committed:
Add CI test and bazel  test using new driver with setup-nvidia action
driver-version: "550.54.15"

nWEIdia committed:
Revert "Disable CUDA 12.4 build for Windows"

This reverts commit 0e23765b99d61fa2fa8378237c7570faeedcb317.

nWEIdia committed:
Revert "Change builder repo/branch to test this PR and https://github.com/pytorch/builder/pull/1786"

This reverts commit a1e3a36a695618aa4cf6a044103e05c963443415.

nWEIdia committed:
Running ./regenerate.sh again after reverting the disabling of Windows
CU124

albanD commented:
Skipping sanity check because this is generated files.

johnnynunez commented:
Is it ready?😍

nWEIdia commented:
> Is it ready?😍

Almost there! Hopefully sometime next week~ Thanks for keeping an eye on this!

nWEIdia committed:
Reverting driver version changes as the default should already be
550.54.15 since
    https://github.com/pytorch/test-infra/commit/d5695df7271be0bfac6090058fb2980ab3052d3f
is in main branch of pytorch/test-infra

nWEIdia commented:
@pytorchbot merge

nWEIdia commented:
@pytorchbot merge

atalman commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -360,12 +374,13 @@ def generate_wheels_matrix(
                         ),
                         "pytorch_extra_install_requirements":
-                        PYTORCH_EXTRA_INSTALL_REQUIREMENTS["12.1"]  # fmt: skip
+                        PYTORCH_EXTRA_INSTALL_REQUIREMENTS["12.4"]  # fmt: skip
```
```suggestion
                        PYTORCH_EXTRA_INSTALL_REQUIREMENTS["12.1"]  # fmt: skip
```
We need to keep 12.1 here still since this is the wheel we will be uploading to pypi 12.4 should still be experimental build for now.

nWEIdia committed:
Update .github/scripts/generate_binary_build_matrix.py

We need to keep 12.1 here still since this is the default wheel we will be uploading to pypi 12.4 should still be experimental build for now.

Co-authored-by: Andrey Talman <atalman@fb.com>

nWEIdia committed:
Run ./regenerate after 12.4 -> 12.1 change in PYTORCH_EXTRA_INSTALL_REQUIREMENTS

johnnynunez commented:
Is it 12.4.1 or 12.4?
https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html

nWEIdia commented:
The cuda used in here is just 12.4 not 12.4.1. The dependency packages used were pulled from 12.4 release page.

nWEIdia committed:
Run .github/scripts/generate_ci_workflows.py to pass workflow checks!

atalman commented:
@pytorchmergebot merge -f "All required jobs are passing"
